### PR TITLE
Implement Quota support in Google service control adapater

### PR DESCRIPTION
**Pull request by manlinl**
Implement Quota support in Google service control adapater

### CODE REVIEW

File: mixer/adapter/svcctrl/quotaprocessor.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+		AllocateOperation: op,
+	}, nil
+}
+
+func getAllocatedQuotaAmount(args *adapter.QuotaArgs, quotaMetrics []*sc.MetricValueSet,
+	quotaCfg *config.Quota) int64 {
+	// Initialize allocated amount to requested amount
+	var amount = args.QuotaAmount
+	for _, metric := range quotaMetrics {
+		if metric.MetricName == quotaAllocationResultMetric &&
```
Comment (apicl) [inline mixer/adapter/svcctrl/quotaprocessor.go]: Is it worth to build a reverse map from Google quota name to mixer quota name?

File: mixer/adapter/svcctrl/quotaprocessor.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+					amount = *value.Int64Value
+					break
+				}
+			}
+		}
+	}
+	return amount
+}
+
+func (p *quotaImpl) ProcessQuota(ctx context.Context,
```
Comment (apicl) [inline mixer/adapter/svcctrl/quotaprocessor.go]: Please add comments to functions.

File: mixer/adapter/svcctrl/quotaprocessor_test.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+			},
+		},
+	}
+}
+
+func TestProcessQuota(t *testing.T) {
+	testCases := []*quotaTestCase{
+		{
+			adapter.QuotaArgs{
+				"-1-",
```
Comment (apicl) [inline mixer/adapter/svcctrl/quotaprocessor_test.go]: I found this "-1-" is really difficult to understand. Poor design.

Not your issue, though.

File: mixer/adapter/svcctrl/quotaprocessor.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+				MetricValues: []*sc.MetricValue{
+					{
+						Int64Value: getInt64Address(args.QuotaAmount),
+					},
+				},
+			},
+		},
+		QuotaMode: getQuotaMode(args),
+	}
+
```
[the remark this answers is not in the document]
Comment (manlinl) [reply mixer/adapter/svcctrl/quotaprocessor.go]: done

File: mixer/adapter/svcctrl/utils_test.go [hunk 5]
```diff
@@ -137,6 +153,29 @@ func TestGenerateConsumerIDFromAPIKey(t *testing.T) {
...
 
+func TestDimensionToString(t *testing.T) {
+	var dim = map[string]interface{}{
+		"key1": "abc",
+		"key2": 123,
+	}
+
+	str, ok := dimensionToString(dim, "key1")
+	if !ok || str != "abc" {
+		t.Errorf(`expect dimensionToString(dim, "key1") return "abc",true, but get %v, %v`,
```
[the remark this answers is not in the document]
Comment (manlinl) [reply mixer/adapter/svcctrl/utils_test.go]: done

File: mixer/adapter/svcctrl/quotaprocessor.go (hunk 1 above)
Comment (manlinl) [reply mixer/adapter/svcctrl/quotaprocessor.go]: Since today, we handle one quota at a time, it doens't help improve efficiency.

**Pull request merged**